File: ios/bridge/ExpoBridgeCatalog.swift
import Foundation

@available(iOS 16.0, *)
enum ExpoBridgeCatalog {
  static func catalogSearch(
    service: CatalogService,
    term: String,
    types: [String],
    options: NSDictionary
  ) async throws -> [String: Any] {
    let pagination = BridgePagination(from: options)
    let searchOptions = CatalogService.SearchOptions(limit: pagination.limit, offset: pagination.offset)
    let result = try await service.search(term: term, types: types, options: searchOptions)
    return BridgeResponses.catalogSearch(result)
  }

  static func getCatalogAlbumTracks(
    service: CatalogService,
    albumId: String,
    options: NSDictionary
  ) async throws -> [String: Any] {
    let pagination = BridgePagination(from: options)
    let searchOptions = CatalogService.SearchOptions(limit: pagination.limit, offset: pagination.offset)
    let songs = try await service.getAlbumTracks(albumId: albumId, options: searchOptions)
    return BridgeResponses.songs(songs)
  }

  static func getCatalogArtistAlbums(
    service: CatalogService,
    artistId: String,
    options: NSDictionary
  ) async throws -> [String: Any] {
    let pagination = BridgePagination(from: options)
    let searchOptions = CatalogService.SearchOptions(limit: pagination.limit, offset: pagination.offset)
    let albums = try await service.getArtistAlbums(artistId: artistId, options: searchOptions)
    return BridgeResponses.albums(albums)
  }

  static func getCatalogPlaylistTracks(
    service: CatalogService,
    playlistId: String,
    options: NSDictionary
  ) async throws -> [String: Any] {
    let pagination = BridgePagination(from: options)
    let searchOptions = CatalogService.SearchOptions(limit: pagination.limit, offset: pagination.offset)
    let songs = try await service.getPlaylistTracks(playlistId: playlistId, options: searchOptions)
    return BridgeResponses.songs(songs)
  }

  static func getCatalogCharts(
    service: CatalogService,
    types: [String],
    options: [String: Any]
  ) async throws -> [String: Any] {
    let pagination = BridgePagination(from: options as NSDictionary)
    let searchOptions = CatalogService.SearchOptions(limit: pagination.limit, offset: pagination.offset)
    let genre = options["genre"] as? String
    let chart = options["chart"] as? String
    let result = try await service.getCharts(types: types, options: searchOptions, genre: genre, chart: chart)
    return BridgeResponses.catalogCharts(result)
  }
}

File: android/src/main/java/expo/modules/applemusic/bridge/ExpoBridgeCatalog.kt
package expo.modules.applemusic.bridge

import expo.modules.applemusic.AndroidCatalogService
import expo.modules.applemusic.BridgeResponses
import expo.modules.applemusic.PaginationOptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.ModuleDefinitionBuilder

internal fun ModuleDefinitionBuilder.registerCatalogBridge(
  catalogService: () -> AndroidCatalogService,
) {
  AsyncFunction("catalogSearch") Coroutine { term: String, types: List<String>, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.catalogSearch(catalogService().search(term, types, pagination))
  }

  AsyncFunction("getCatalogSong") Coroutine { id: String ->
    catalogService().getSong(id)
  }

  AsyncFunction("getCatalogAlbum") Coroutine { id: String ->
    catalogService().getAlbum(id)
  }

  AsyncFunction("getCatalogArtist") Coroutine { id: String ->
    catalogService().getArtist(id)
  }

  AsyncFunction("getCatalogPlaylist") Coroutine { id: String ->
    catalogService().getPlaylist(id)
  }

  AsyncFunction("getCatalogStation") Coroutine { id: String ->
    catalogService().getStation(id)
  }

  AsyncFunction("getCatalogMusicVideo") Coroutine { id: String ->
    catalogService().getMusicVideo(id)
  }

  AsyncFunction("getCatalogAlbumTracks") Coroutine { albumId: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.songs(catalogService().getAlbumTracks(albumId, pagination))
  }

  AsyncFunction("getCatalogArtistAlbums") Coroutine { artistId: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.albums(catalogService().getArtistAlbums(artistId, pagination))
  }

  AsyncFunction("getCatalogPlaylistTracks") Coroutine { playlistId: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.songs(catalogService().getPlaylistTracks(playlistId, pagination))
  }

  AsyncFunction("getCatalogCharts") Coroutine { types: List<String>, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    val genre = options["genre"] as? String
    val chart = options["chart"] as? String
    BridgeResponses.catalogCharts(catalogService().getCharts(types, pagination, genre, chart))
  }
}
